### DISCUSSION TITLE
Make an actual Java installation tutorial

### DISCUSSION THREAD

**Discussion by KaKi87** [tolgee/tolgee-platform#1024] [Ideas]
Hello,

Running Tolgee using Java isn't as simple as `java -jar tolgee-<version>.jar --server.port=""` like the [documentation](https://tolgee.io/docs/platform/self_hosting/running_with_java) says.

Here's what I had to do to make it work, on Debian 10 through an Apache proxy.

## Java version

First, the `default-jre` package isn't enough, because the first error I got was similar to the following :
> [...] has been compiled by a more recent version of the Java Runtime (class file version [...]), this version of the Java Runtime only recognizes class file versions up to [...]

So, I had to install a newer version, like this :
```bash
sudo add-apt-repository ppa:linuxuprising/java
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys EA8CACC073C3DB2A
sudo apt update
sudo apt install oracle-java17-installer --install-recommends
```
So that `java -version` would output this : `java version "17.0.1" 2021-10-19 LTS`

## Database configuration

Then, I realized that making the program use my own PostgreSQL wasn't only about adding `tolgee.postgres-autostart.enabled=false`, but also specifying the following configuration items :
- `spring.datasource.url`
- `spring.datasource.username`
- `spring.datasource.password`

One could say that specifying the database's configuration is an obvious step, but I could have expected it to occur from the web UI, like most self-hostable web-based projects maintained by a company do.

I also experienced the following error :
> permission denied to create extension "pg_trgm"

So, I had to the following in `psql` : `ALTER ROLE tolgee CREATEROLE SuperUser;`
Which I reverted once the platform was running : `ALTER USER tolgee NOSUPERUSER;`
Because, obviously, an app shouldn't have full control on a database that multiple apps (and maybe users) share.

Ideally, it would be nice not to need it at all.

## WebSocket port

Once the app is successfully running, logs show a Socket.IO server running on a different port than the HTTP port.
But, when you want to specify the HTTP port, you might also want to specify the WebSocket port.
So, the following configuration item must be added : `tolgee.socket-io.port`

## Additional configuration

While browsing the [server configuration](https://tolgee.io/docs/platform/self_hosting/configuration) page, I also added `tolgee.front-end-url`, because it felt useful, whether using Docker or not, as well as `tolgee.authentication.registrations-allowed=true` since the instance is meant to be public.

For the same reason, I also added `tolgee.recaptcha.*` and `tolgee.smtp.*` settings, but those are user-specific choices I gess.

## Result

Considering the number of configuration items to put, I created a `start.sh` script that looks like this :
```bash
#!/usr/bin/env bash
java -jar ./tolgee-v2.13.0.jar \
    --server.port=0000 \
    --spring.datasource.url="jdbc:postgresql://localhost:5432/tolgee" \
    --spring.datasource.username="" \
    --spring.datasource.password="" \
    --tolgee.front-end-url="https://i18n.kaki87.net" \
    --tolgee.postgres-autostart.enabled=false \
    --tolgee.authentication.enabled=true \
    --tolgee.authentication.registrations-allowed=true \
    --tolgee.recaptcha.site-key="" \
    --tolgee.recaptcha.secret-key="" \
    --tolgee.smtp.host="smtp.eu.mailgun.org" \
    --tolgee.smtp.username="postmaster@mail.kaki87.net" \
    --tolgee.smtp.password="" \
    --tolgee.smtp.port=0000 \
    --tolgee.smtp.auth=true \
    --tolgee.smtp.ssl-enabled=true \
    --tolgee.smtp.from="Tolgee <postmaster@mail.kaki87.net>" \
    --tolgee.socket-io.port=0000
```

Perhaps I'll encounter other problems that could have been avoided if I knew all the configuration steps from the beginning, I'll keep you informed here.

Meanwhile, it would be nice to make an actual tutorial for self-hosting Tolgee without Docker.

Thanks

**Comment by JanCizmar** [comment]
Thanks for opening this discussion and sharing your experience. This could help feature travellers a lot!

Let me react to some of your points. 
> First, the default-jre package isn't enough

This is odd. It seems like default version of java in debian 10 is 11 and Tolgee is built with Java 11. What version Java had you installed before installing v17?

> Then, I realized that making the program use my own PostgreSQL wasn't only about adding tolgee.postgres-autostart.enabled=false, but also specifying the following configuration items. 

Hmm, yep, but its described in [configuraion section.](https://tolgee.io/docs/platform/self_hosting/configuration#data-source-settings)

> Once the app is successfully running, logs show a Socket.IO server running on a different port than the HTTP port.

Websockets are not currently used on frontend. We will replace SocketIO with Spring's native approach. So there is currently no need to expose the port.

> Meanwhile, it would be nice to make an actual tutorial for self-hosting Tolgee without Docker.

Yep, you're right. I'll write a blog post about self hosting this or the next week.

Thanks a lot for sharing again.

**Reply by KaKi87** [reply]
> Thanks for opening this discussion and sharing your experience. This could help feature travellers a lot!

You're welcome, they're welcome !
> This is odd. It seems like default version of java in debian 10 is 11 and Tolgee is built with Java 11.

Indeed, I'll look into it.
> Hmm, yep, but its described in configuraion section.

Yes, but if there were a tutorial, every user wouldn't need to go through the entire the configuration reference making sure they got everything they would need.

For most users, the configuration will pretty much look just like mine.

> Websockets are not currently used on frontend.

I noticed it too !
> there is currently no need to expose the port

So, should I use `tolgee.socket-io.enabled=false` ?
> I'll write a blog post about self hosting this or the next week.

👍

**Reply by JanCizmar** [reply]
> So, should I use tolgee.socket-io.enabled=false ?

You can.

**Comment by JanCizmar** [comment]
I also wonder if you would be interested into writing an article about your experience. We would be glad to share it on our blog! :)

**Reply by KaKi87** [reply]
Maybe in a few days or weeks, when I'll have fully tested the platform. I'm also working on a lib to allow two-way JSON syncing from Node environments.
